indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java

```diff
@@ -65,30 +66,33 @@ public MergeTask(
     this.aggregators = Preconditions.checkNotNull(aggregators, "null aggregations");
     this.rollup = rollup == null ? Boolean.TRUE : rollup;
     this.indexSpec = indexSpec == null ? new IndexSpec() : indexSpec;
+    this.customIndexMerger = customIndexMerger;
   }
 
   @Override
   public File merge(final TaskToolbox toolbox, final Map<DataSegment, File> segments, final File outDir)
       throws Exception
   {
-    IndexMerger indexMerger = toolbox.getIndexMergerV9();
-    return indexMerger.mergeQueryableIndex(
-        Lists.transform(
-            ImmutableList.copyOf(segments.values()),
-            new Function<File, QueryableIndex>()
+    IndexMerger indexMerger = customIndexMerger != null ? customIndexMerger : toolbox.getIndexMergerV9();
+    FunctionalIterable<QueryableIndex> indexes = FunctionalIterable
+        .create(segments.entrySet())
+        .transform(
+            new Function<Map.Entry<DataSegment, File>, QueryableIndex>()
             {
               @Override
-              public QueryableIndex apply(@Nullable File input)
+              public QueryableIndex apply(Map.Entry<DataSegment, File> input)
               {
                 try {
-                  return toolbox.getIndexIO().loadIndex(input);
+                  return toolbox.getIndexIO().loadSegment(input.getKey(), input.getValue()).asQueryableIndex();
```

Comment:
every place want to load index from file should call loadSegment instead of loadIndex. inside `loadSegment`, it will use the customized segmentizerFactory to map the file to index.
Most usages of loadIndex() are in tests and benchmarks, so I don't modify them.
I will double check the rest usages to make sure every place calling loadIndex is replaced by loadSegment()